File: npm/src/commands/validate.js
'use strict'
const { Command } = require('commander')
const { validate, isLenient, lenientUntilDate, getItemMeta } = require('../validator')
const { t } = require('../i18n')

const cmd = new Command('validate')
cmd.description(t('validate.description'))
cmd.option('--json', 'output result as JSON')
cmd.action(async (options) => {
  const { violations, warnings } = await validate()
  const lenient = isLenient()
  const mode = lenient ? 'lenient' : 'strict'
  const exitCode = violations.length > 0 ? 1 : 0

  if (options.json) {
    const output = {
      summary: {
        violations: violations.length,
        warnings: warnings.length,
        mode,
        exit_code: exitCode,
      },
      violations: violations.map(v => { const m = getItemMeta(v); return { message: v, rule: m.rule, file: m.file } }),
      warnings: warnings.map(w => { const m = getItemMeta(w); return { message: w, rule: m.rule, file: m.file } }),
    }
    console.log(JSON.stringify(output, null, 2))
    process.exit(exitCode)
    return
  }

  // Informar usuário sobre modo lenient
  if (lenient) {
    const until = lenientUntilDate()
    if (until) {
      console.log(`[LENIENT MODE] ${t('validate.lenient_mode', { date: until })}`)
    }
  }

  if (violations.length === 0 && warnings.length === 0) {
    console.log(t('validate.ok'))
    return
  }

  if (violations.length > 0) {
    console.log(`\n${t('validate.violations', { count: violations.length })}`)
    violations.forEach(v => console.log(`  • ${v}`))
  }

  if (warnings.length > 0) {
    console.log(`\n${t('validate.warnings', { count: warnings.length })}`)
    warnings.forEach(w => console.log(`  • ${w}`))
  }

  if (violations.length > 0) process.exit(1)
})

module.exports = cmd

File: internal/commands/validate.go
package commands

import (
	"encoding/json"
	"fmt"
	"strconv"

	"github.com/kgsaran/trackfw/internal/i18n"
	"github.com/kgsaran/trackfw/internal/validator"
	"github.com/spf13/cobra"
)

func newValidateCmd() *cobra.Command {
	var jsonOutput bool

	cmd := &cobra.Command{
		Use:   "validate",
		Short: i18n.T("validate.description"),
		RunE: func(cmd *cobra.Command, args []string) error {
			if jsonOutput {
				// Modo JSON: usa ValidateTagged para propagar rule+file nos RuleItems.
				cmd.SilenceErrors = true
				cmd.SilenceUsage = true

				taggedV, taggedW, err := validator.ValidateTagged()
				if err != nil {
					return err
				}

				result := validator.BuildResultTagged(taggedV, taggedW, validator.IsLenient())
				data, marshalErr := json.Marshal(result)
				if marshalErr != nil {
					return marshalErr
				}
				fmt.Fprintln(cmd.OutOrStdout(), string(data))

				if len(taggedV) > 0 {
					return fmt.Errorf("%s", i18n.T("validate.violations", "count", strconv.Itoa(len(taggedV))))
				}
				return nil
			}

			// Modo texto (comportamento original inalterado).
			violations, warnings, err := validator.Validate()
			if err != nil {
				return err
			}

			if validator.IsLenient() {
				until := validator.LenientUntilDate()
				if until != "" {
					fmt.Printf("[LENIENT MODE] %s\n", i18n.T("validate.lenient_mode", "date", until))
				}
			}
			for _, w := range warnings {
				fmt.Printf("⚠  %s\n", w)
			}
			if len(violations) == 0 {
				if len(warnings) == 0 {
					fmt.Println(i18n.T("validate.ok"))
				} else {
					fmt.Println(i18n.T("validate.warnings", "count", strconv.Itoa(len(warnings))))
				}
				return nil
			}
			for _, v := range violations {
				fmt.Printf("✗ %s\n", v)
			}
			return fmt.Errorf("%s", i18n.T("validate.violations", "count", strconv.Itoa(len(violations))))
		},
	}

	cmd.Flags().BoolVar(&jsonOutput, "json", false, "Output validation result as JSON")
	return cmd
}
